### PR TITLE
Add Documentation for Raw-Request-URI synthetic header #759

### PR DESCRIPTION
**Pull request by jlprat**
Issue: #759
Contains:
* Java documentation
* Cleanup in comments for ModelDocTest
* Move Scala example to its own class and use @@snip instead of inline code

### CODE REVIEW

**Comment by jonas** [comment]
Looks like there is a `sed` command which needs a check whether the input file exists
```
+ sed -i 's/<failure /<skipped /g' ./akka-http2-support/target/test-reports/h2spec-junit.xml
sed: can't read ./akka-http2-support/target/test-reports/h2spec-junit.xml: No such file or directory
Build step 'Execute a set of scripts' changed build result to FAILURE
```

File: docs/src/main/paradox/java/http/common/http-model.md [hunk 1]
```diff
@@ -53,6 +53,23 @@ new immutable instance, so instances can be shared freely. There exist some over
 simplify creating requests for common cases. Also, to aid readability, there are predefined alternatives for `create`
 named after HTTP methods to create a request with a given method and URI directly.
 
+<a id="synthetic-headers-scala"></a>
+### Synthetic Headers
+
+In some cases it may be necessary to deviate from fully RFC-Compliant behavior. For instance, Amazon S3 treats 
+the + character in the path part of the URL as a string, even though the RFC specifies that this behavior should
```
Comment (jonas) [inline docs/src/main/paradox/java/http/common/http-model.md]: Rereading #451, I think we should fix the Java and Scala docs to say:

```
Amazon S3 treats the `+`  character in the path part of the URL as a space
```

File: docs/src/main/paradox/java/http/common/http-model.md [hunk 2]
```diff
@@ -53,6 +53,23 @@ new immutable instance, so instances can be shared freely. There exist some over
...
+be limited exclusively to the query portion of the URI.
+
+In order to work around these types of edge cases, Akka HTTP provides for the ability to provide extra, 
+non-standard information to the request via synthetic headers. These headers are not passed to the client
+but are instead consumed by the request engine and used to override default behavior.
+
+For instance, in order to provide a raw request uri, bypassing the default url normalization, you could do the
+following:
+
+@@snip [ModelDocTest.java](../../../../../test/java/docs/http/javadsl/ModelDocTest.java) { #synthetic-header-s3 }
```
Comment (jonas) [inline docs/src/main/paradox/java/http/common/http-model.md]: It is possible to also have imports in the Java examples by using two identifiers unfortunately Paradox will not normalize the indentation (yet!).

```java
//#import-raw-request-uri
import akka.http.javadsl.model.headers.RawRequestURI
//#import-raw-request-uri
...
//#synthetic-header-s3
...
//#synthetic-header-s3
```
and then
```md
@@snip [ModelDocTest.java](../../../../../test/java/docs/http/javadsl/ModelDocTest.java) { #import-raw-request-uri #synthetic-header-s3 }
```

**Comment by jrudolph** [comment]
PLS BUILD

File: docs/src/main/paradox/java/http/common/http-model.md (hunk 1 above)
Comment (jlprat) [inline docs/src/main/paradox/java/http/common/http-model.md]: Agree, sounds better

File: docs/src/main/paradox/java/http/common/http-model.md (hunk 2 above)
Comment (jonas) [inline docs/src/main/paradox/java/http/common/http-model.md]: Makes sense

**Comment by jonas** [comment]
Great. Merged by hand to also fix the `<a id=...>` I just noticed.

**Comment by jrudolph** [comment]
Thanks, @jlprat and @jonas.

**Pull request closed without merging**